skimage/future/objdetect/cascade.pyx

```diff
@@ -0,0 +1,225 @@
+# cython: cdivision=True
+# cython: boundscheck=False
+# cython: nonecheck=False
+# cython: wraparound=False
+
+import numpy as np
+cimport numpy as cnp
+from libc.stdlib cimport malloc, free
+import xml.etree.ElementTree as ET
+from ...feature._texture cimport _multiblock_lbp
+
+cdef struct MBLBP:
+
+        Py_ssize_t r
+        Py_ssize_t c
+        Py_ssize_t width
+        Py_ssize_t height
+
+cdef struct MBLBPStump:
+
+        Py_ssize_t feature_id
+        Py_ssize_t lut_idx
+        float left
+        float right
+
+cdef struct Stage:
+
+        Py_ssize_t first_idx
+        Py_ssize_t amount
+        float threshold
+
+cdef class Cascade:
+
+    cdef:
+        public float eps
+        public Py_ssize_t stages_amount
+        public Py_ssize_t stumps_amount
+        public Py_ssize_t features_amount
+        public Py_ssize_t window_width
+        public Py_ssize_t window_height
+        Stage * stages
+        MBLBPStump * stumps
+        MBLBP * features
+        cnp.uint32_t * LUTs
+
+    def __dealloc__(self):
+
+        # Free the memory that was used for c-arrays.
+        free(self.stages)
+        free(self.stumps)
+        free(self.features)
+        free(self.LUTs)
+
+    def evaluate(self, float[:, ::1] int_img):
+
+        cdef:
+            float stage_threshold
+            float stage_points
+            int lbp_code
+            int bit
+            Py_ssize_t stage_number
+            Py_ssize_t weak_classifier_number
+            Py_ssize_t feature_number
+            Py_ssize_t features_amount
+            Py_ssize_t stumps_amount
+            Py_ssize_t first_stump_idx
+            Py_ssize_t lut_idx
+            cnp.uint32_t[::1] current_lut
+            Stage current_stage
+            MBLBPStump current_stump
+            MBLBP current_feature
+
+
+        for stage_number in range(self.stages_amount):
+
+            current_stage = self.stages[stage_number]
+            first_stump_idx = current_stage.first_idx
+            stage_points = 0
+
+            for weak_classifier_number in range(current_stage.amount):
+
+                current_stump = self.stumps[first_stump_idx + weak_classifier_number]
+
+                current_feature = self.features[current_stump.feature_id]
+
+                lbp_code = _multiblock_lbp(int_img,
+                                           current_feature.r,
+                                           current_feature.c,
+                                           current_feature.width,
+                                           current_feature.height)
+
+                lut_idx = current_stump.lut_idx
+
+                bit = (self.LUTs[lut_idx + (lbp_code >> 5)] >> (lbp_code & 31)) & 1
+
+                stage_points += current_stump.left if bit else current_stump.right
+
+            if stage_points < (current_stage.threshold - self.eps):
+                return False
+
+        return True
+
+    def load_xml(self, filename, eps=1e-5):
```

Comment:
This function is enormous. There's a rule of thumb that functions should not be more than 25 lines long. This isn't a hard rule, but there are good reasons for it.

Candidates for abstraction are every point where you commented a chunk of code. Replace these chunks with well-named functions.

This comment applies to other parts of this PR, too.